### PR TITLE
Add the option to turn on async-TP

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #429

This PR adds the option to turn on async-TP (`--experimental.enable_async_tensor_parallel`). The feature is currently implemented as compiler passes on relevant patterns, so the option is currently only effective when compile is enabled.

Some trace samples from llama3_70b with tp degree=8:

**all-gather -> qkv projection**
Baseline:
image

With async-TP:
image

**ffn -> reduce-scater**
Baseline:
image

With async-TP:
image

**all-gather -> ffn**
Baseline:
image

With async-TP:
image

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -252,6 +252,11 @@ def loss_fn(pred, labels):
         for m in model_parts
     ]
 
+    # for ease of testing TP in lieu of FSDP
+    if job_config.training.tensor_parallel_degree == world_size:
+        for model in model_parts:
+            model.to(torch.bfloat16)
```
Comment (tianyu-l) [inline train.py]: May I ask what are we doing here?

Comment (yifuwang) [reply train.py]: Hey @tianyu-l, this is to allow conveniently testing `tp_degree=8` in bfloat16 on a single host. When running with `tp_degree=8` on a single 8-GPU host, FSDP will not be used and the model will be in float32. I figured since we don't care about float32, we might as well move the model to bfloat16 when FSDP is not used.

I can remove this if we feel it is janky. This logic is not related to async-TP.

Comment (tianyu-l) [reply train.py]: oh I see... I thought we still care about fp32 in TP because bf16 is lossy when doing reductions. Maybe I'm lacking some context.

Even if we would offer precision adjustment for TP, I feel we should have an option for user to decide as in FSDP, like [here](https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L310).

**Comment by tianyu-l** [comment]
another question brought up by @awgu offline:
Currently we are doing Sequence Parallel with activations sharded on the sequence dimension `Shard(1)`, rather than folding the first two dimensions and do `Shard(0)`. Would this affect correctness of async TP?
I wonder if this has been tested, e.g. after disabling randomness we can observe the same loss curve for TP and async TP.

File: train.py (hunk 1 above)
Comment (yifuwang) [reply train.py]: Sounds good. I'll remove this for now and explore other options later

**Comment by yifuwang** [comment]
@tianyu-l yeah this scenario has been tested and I've verified the loss curve with 64 GPUs at 1000 steps: https://docs.google.com/spreadsheets/d/1jt35XG3RzCEDyEZ3TD5bpCQ-760UQ0Yv4cUEeUHU8YM/edit?gid=0#gid=0

Currently async-TP is agnostic of DTensor. It is done by performing semantically equivalent rewrite on the desugared aten IR graph. I don't think anything sharding related can go wrong with this approach.

**Review by tianyu-l — APPROVED** [review]
sgtm! Thanks for the great work!
yeah let's discuss dtype thing later

**Comment by awgu** [comment]
Run ghstack land <PR url> I think

**Comment by yifuwang** [comment]
Thanks @awgu!

**Review by fduwjj — COMMENTED** [review]
Nice!!
